bson/src/main/org/bson/codecs/configuration/mapper/ClassModel.java

```diff
@@ -91,4 +81,72 @@ public FieldModel getField(final String name) {
     public List<FieldModel> getFields() {
         return new ArrayList<FieldModel>(fields.values());
     }
+
+    /**
+     * Returns the name of the class represented by this ClassModel
+     *
+     * @return the name
+     */
+    public String getName() {
+        return getType().getName();
+    }
+
+    /**
+     * Suggests a new value for the collection name.
+     *
+     * @param weight         The weight to give this value relative to other values set
+     * @param collectionName the new collection name to suggest
+     * @see WeightedValue
+     */
+    public void setCollectionName(final Integer weight, final String collectionName) {
+        this.collectionName.set(weight, collectionName);
+    }
+
+    /**
+     * Executes the actual mapping of the class.
+     */
+    public void map() {
+        if (!mapped) {
+            final ResolvedType resolved = resolver.resolve(getType());
+            final ResolvedTypeWithMembers type = memberResolver.resolve(resolved, null, null);
+
+            for (final ResolvedType resolvedType : resolved.getTypeParameters()) {
+                addParameter(resolvedType.getErasedType());
+            }
+
+            for (final ResolvedField field : type.getMemberFields()) {
```

Comment:
I realized that by doing this, i can effectively hide all the reflection/introspection away.